nexus/src/app/mod.rs

```diff
@@ -173,8 +174,17 @@ impl Nexus {
                 "component" => "DpdClient"
             )),
         };
-        let (dpd_host, dpd_port) = if let Some(dpd_address) =
-            &config.pkg.dendrite.address
+
+        // TODO
```

Comment:
I honestly don't know. I actually discovered another problem with this logic last night during multi-switch testing on physical machines, so this entire block of code may end up changing / going away as a whole.

When multiple switch zones are registered in DNS, they're registered under the same namespace, so resolution is unpredictable (you might get the address for switch0 or switch1). Looks like we will need to go ahead and plumb in some additional logic for selecting a specific switch, which adds its own complexity. The problem I ran into last night was the nat entries for nexus and other internal services were created on the wrong switch.

A somewhat half baked idea: is it harmful to put the nat entries on both switches by default? The initial thought is that outbound traffic will only be routed to a switch with the boundary services addr, so (possibly?) no harm there, inbound traffic will always arrive at the correct location because both switches should know how to reach all the sleds. Thoughts?

Reply:
I think the DNS setup here is flexible - we probably should have unique identifiers for each service, so we can query for "rack X, switch 0" or "rack X, switch 1". If each service is distinct, DNS also should not be treating the services as interchangeable.

Reply:
@smklein currently the DNS entries are created before we actually know which dendrite service manages which switch. I haven't dug in enough to know if this can be deferred, but after the conversation @rcgoodfellow and I just had (notes incoming!) it may be moot anyway, as I think we've figured out an approach that solves this problem without relying on DNS.

Reply:
Levon and I had a live discussion about this. The determinations were the following.

1. Only place NAT entries on switches that have external routing configured. There is no purpose for NAT entries on switches that have no external path. Because NAT entries are specific to the ingress address from the rack's perspective, there is not a more specific discriminator that applies. As we evolve toward the internet gateway concept in RFD 21 we may have more specific discriminators available and can be more specific about what VPCs have access to what external routes, but that day is not today.

2. When making decisions in the control plane about what configuration needs to go on what switches there are two basic scenarios. a) we are in the context of an operator API or RSS call where the switch has been explicitly specified, easy - the configuration goes to the specified switch. b) we are configuring switches as a part of an operator or user API call, or internal automation, that does not call out a specific switch - like setting up NAT for an instance or internal service. In this case, we take a property-based approach to figure out what switches the configuration applies to. Item (1) above is an example of this approach. When setting up NAT, find the set of switches that have external routing enabled, and add the NAT entries to those switches.

Reply:
This has now been implemented for RSS. Adding requisite properties needed for implementing logic in Nexus.

Reply:
> Just to confirm, when you say external path you mean specifically for the external IP we'll be NATing?

By external path, I mean some set of routes out the front ports of a sidecar to some set of destinations on the upstream network. Because we do source NAT, the particular IPs we NAT are not strictly related to the routes on the sidecar. For example, an administrator may only set up routes to `10.10.0.0/16`, but assign addresses in `10.20.0.0/16` to the rack. In that case, we'll NAT instance and service addresses onto `10.20.0.0/16`. OPTE and the sidecar with the route to `10.10.0.0/16` then do their cooperative NAT thing to transit packets between `10.20.0.0/16`/`10.10.0.0/16`.

On a sidecar with no routes, egress just isn't possible. I suppose ingress is, as we could just get packets from the upstream network without any routes, look at the destination address, and if it's in the NAT tables, encap it and send it down to the target sled. That's a bit of an odd asymmetric setup having one sidecar be strictly ingress. But I guess if we think about the rack more as a holistic computer and not a rack of computers with two switches, perhaps it makes sense. However, broadening that logic to a cluster of racks, we definitely don't want to populate all NAT rules on all racks if only some racks have routes to upstream networks.

So I guess this is a long-winded way of saying: considering the rack as a whole (both sidecars) as a unit in terms of NAT provisioning is not unreasonable.